core/src/main/scala/cats/data/Chain.scala

```diff
@@ -617,19 +626,57 @@ sealed abstract class Chain[+A] {
 
   /**
    * Remove duplicates. Duplicates are checked using `Order[_]` instance.
+   * 
+   * Example:
+   * {{{
+   * scala> import cats.data.Chain
+   * scala> val chain = Chain(1, 2, 2, 3)
+   * scala> chain.distinct
+   * res0: cats.data.Chain[Int] = Chain(1, 2, 3)
+   * }}}
    */
   def distinct[AA >: A](implicit O: Order[AA]): Chain[AA] = {
-    implicit val ord: Ordering[AA] = O.toOrdering
-
-    var alreadyIn = TreeSet.empty[AA]
+    if (isEmptyOrSingleton) this
+    else {
+      implicit val ord: Ordering[AA] = O.toOrdering
+
+      var result = Chain.empty[A]
+      val seen = mutable.TreeSet.empty[AA]
+      val it = iterator
+      while (it.hasNext) {
+        val next = it.next()
+        if (seen.add(next))
+          result ++= one(next)
```

Comment:
why `result ++= one(next)` vs `result = result :+ next`

Why not use the method we already have?

Reply:
Well, `chain :+ a` does the same thing as `chain ++ one(a)`.
I.e. we cannot avoid creating a singleton here anyway.
But .. I agree – let's switch to more explicit method `:+` here.

Reply:
fixed